src/ucp/core/ucp_request.c

```diff
@@ -415,6 +415,11 @@ void ucp_request_send_state_ff(ucp_request_t *req, ucs_status_t status)
                   ucs_status_string(status));
 
     if (req->send.uct.func == ucp_proto_progress_am_single) {
+        ucs_assert(req->send.state.uct_comp.count == 0);
+
+        if (req->id != UCP_REQUEST_ID_INVALID) {
+            ucp_request_id_release(req);
+        }
```

Comment:
maybe do it for all requests? e.g before/after this if

Reply:
We can’t do it in a common code. No need to check completion counter for requests which don’t set completion callback.

Reply:
do you mean we will release id which was not allocated? or later on the completion function will try to release an invalid id?

Reply:
No, I meant that some of requests don’t have uct_comp::count set to 0 (i.e. requests which don’t set completion callback - e.g. AM a copy) when we do fast-forward. So l, we can’t touch counter to check whether we need to call completion callback or not.

Reply:
but, can we do ` if (req->id != UCP_REQUEST_ID_INVALID) { ucp_request_id_release }` in any case?

Reply:
Unfortunately, no.
Since we should do it of completion counter is 0

Reply:
why so? it's not like we're going to send or receive anything on this request?

Reply:
yes, but we should complete the request if it reaches comp_count==0 to avoid use after release

Reply:
@yosefe sorry if I was unclear.
I meant that UCP reset state for each request in `ucp_request_send_state_reset()`, but only for requests that use UCT completion, UCP resets completion counter to `0`.
https://github.com/openucx/ucx/blob/5b5a98a522ce91171c787a4c158ebb7f45703b96/src/ucp/core/ucp_request.inl#L398

For some operations in `ucp_request_send_state_ff()` we should check if completion counter reaches `0` and then complete them. So, we have to identify what the protocol is used by checking its completion callback and then complete the request if completion counter is 0.

Reply:
per offline discussion, make ucp_request_id_release() work even if req->id==UCP_REQUEST_ID_INVALID (as nop)